feature/gnoi/os/tests/osinstall/README.md

```diff
@@ -2,81 +2,89 @@
+
+      b. Transfer the content by issuing gnoi.os.Install rpc with InstallRequest.transfer_content message.
+         * Expect it to return InstallResponse with a TransferProgress status
+         asynchronously at certain intervals.
+         * TODO: When the expected amount of bytes_received is reported by the
+         switch, move to the activation step next.
+
+      c. End the transfer of software by issuing gnoi.os.Install rpc with InstallRequest.TransferEnd message.
+         * Expect the switch to return InstallResponse with a Validated message. The version in the message should be set to the one which was transferred above.
+      
+      d. Activate the software by issuing gnoi.os.Activate rpc.
+         * Set the version field of the ActivateRequest message to be the same as the version specified in the TransferRequest message above.
+         * Set the no_reboot flag to true.
+         * Set the standby_supervisor to FALSE.
+           
+   2. Install and activate the same new software version on standby supervisor:
+      
+      a. Repeat the above process of TransferRequest. This time set the standby_supervisor to TRUE.
+         * Expect the switch to return a InstallResponse with a SyncProgress message. The switch should sync the software image from primary SUP to standby.
+         * Expect the sync to return a value of 100 for percentage_transferred field.
+         * At the end, expect the switch to return InstallResponse with a Validated message. The version in the message should be set to the one which was transferred above.
+           
+   3. Activate the software by issuing gnoi.os.Activate rpc as in the case of primary supervisor.
+      
+      a. Set the version field of the ActivateRequest message to be the same as the version specified in the TransferRequest message above.
+      
+      b. Set the no_reboot flag to true.
+      
+      c. Set the standby_supervisor to TRUE this time.
+      
+   4. Reboot the switch:
+      
+      a. Issue gnoi.system.Reboot as specified in [gNOI-3.1: Complete Chassis Reboot](feature/gnoi/tests/complete_chassis_reboot/complete_chassis_reboot_test.md).
+      
+   5. Verify that the supervisor image has moved to the new image:
+       
+      a. Verify that the supervisor has a valid image by issuing gnoi.os.Verify rpc.
+         * Expect a VerifyResponse with the version field set to the version specified in messages above eventually.
+         * Verify the standby supervisor version.
+         * Expect that the VerifyResponse.verify_standby has the same version in messages above.
+
+               
+* gNOI-4.1.2: Configuration push verification post OS change and check health of the software-module that allows configuration.
+  1. Check the health of the software-module component that allows configuration of the router and verify if it is **HEALTHY** using the leaf /components/component[**process that handles configuration of the DUT**]/healthz/state/status/
```

Comment:
What should happen if gnmi get for healthz path is not supported  or the below gnoi healthz calls are not supported 
 but the DUT is actually otherwise healthy?  Should gNOI-4.1.2 pass or fail?

For example: : gNOI-4.1.1 succeeds, and gNOI-4.1.2 steps 2 and 3 can succeed, except for the calls to healthz?)

Seems a likely scenario that a DUT could fully support OS install, but not support healthz.  IMO, we should be able to tell which part is failing

Reply:
`What should happen if gnmi get for healthz path is not supported or the below gnoi healthz calls are not supported
but the DUT is actually otherwise healthy? Should gNOI-4.1.2 pass or fail?`
**Response:** This situation should be handled in the similar fashion as we always do for unsupported paths i.e.
 a. Add a deviation
 b. Raise a Google bug (w/ assignee as the implementation that do not support the feature) to track long-term fix for the deviation
 c. Once the fix is available, remove the deviation.

Example: If I was to write a test for BGP, I would write it such that
1. BGP configuration is Pushed
2. Traffic is being forwarded as expected
3. Check all the relevant state paths for BGP to ensure the implementation supports those paths and the results are accurate to the context for each subtest involved.

I wouldnt just do steps 1 and 2 and write a separate test for all state paths and cover "3" there. Imo it is not right  to call a test pass/fail just because bullet#2 is pass/fail and in production we need the state paths to be accurately streamed as well matching the context of the situation.
Similarly in this test i.e. gNOI-4.1, HealthZ check for the process in context seems important. 
As for unsupported paths, we have the respective implementations raise a deviation like we always do and follow the a-c steps above. I have 2xNotes in the readme to cover for implementations that may have a different behavior. The Deviation should have pretty much similar impact as before for the implementations not supporting HealthZ for Software-module other than the fact that this readme also asks to push some configuration. Sorry for the long response but hopefully it helps explain the thought behind the test and the process to handle it when the support doesnt exist?

`Seems a likely scenario that a DUT could fully support OS install, but not support healthz. IMO, we should be able to tell which part is failing`
**Response:** I see the following chain of events for an implementation that do not support HealthZ
1. They will pass/fail gNOI-4.1.1 the same as before.
2. For gNOI-4.1.2 the deviation will skip healthZ check and the implementation will push a configuration. I see 2 offshoots of this operation. Either,
    a. The implementation allows unsupported configuration to be pushed and its daemon keeps crashing. I am hoping the gNMI.Get() RPC call would fail in which case the test will fail.
    b.  The implementation wouldnt allow gNMI.Set() RPC for the unsupported conifguration in which case the test is a failure.
If the configuration push succeeds and the gNMI.Get() operation is also successful for the same configuration then the test is a pass. Hopefully this make sense?

Reply:
osinstall test should ensure the os install completed successfully.  ie: the new os in installed and running.  Leave it to the  other tests to verify all the other features.    os install affects every process and every path on the entire system.  To fully know it's successful one should run all the tests in the entire featureprofiles repository, which includes healthz feature checks.    

This update does include gNMI set and get which will identify the bug that was experienced.  gNMI set and get are well known, widely supported features that are a baseline requirement for nearly all featureprofiles tests.  Healthz is not.  Let the the healthz tests confirm the healthz feature is supported.     Using the bgp analogy, do we know if BGP works after os_install?  We don't know, so we should check it.   Should that check be done in this test?  No, do it in the BGP tests.

Using deviations should be avoided.  Adding healthz here creates the possibility for additional deviations to be required which are not directly related or required for the feature being tested.   

That said, all the above does not mean including healthz breaks this test.  This test is now a os_install AND healthz test.   While I don't agree this is the minimum, this test can still work.  So I will approve.

Reply:
Thinking this through, the test will probably end up flaky if the daemon is crashing continuously as the reported health will depend on its state in time. I have changed the write-up to only update and read config. This seems a better approach and the test would fail if either config push fails or the GET operation fails.

However the business requirement is also to ensure that the health of the config daemon can be streamed and supported by all implementations. Hence I also have https://github.com/openconfig/featureprofiles/pull/2240 to test the paths in a more stable environment.